### ISSUE TITLE
[TW-527] Dependencies are not released when blocking task is done

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#562]
_Arkady Grudzinsky on 2012-06-19T23:30:42Z says:_ 

 If I mark Task2 to depend on Task1, it falls off the "next" list (see Bug# 957).  When I mark Task1 as "done", dependency is not deleted from Task2, and it still falls off the "next" list.  This is a serious issue as I expect released tasks to "reemerge" at the top of my list when they are unblocked, but they remain deprioritized.  

Scenario:
*  Create task1 and task2 depending on task1

<pre>
task> add task1 pro:asdf
Created task 153.
The project 'asdf' has changed.  Project 'asdf' is 50% complete (1 of 2 tasks remaining).

task> add task2 pro:asdf dep:153
Created task 154.
The project 'asdf' has changed.  Project 'asdf' is 33% complete (2 of 3 tasks remaining).

task> list pro:asdf

ID  Dep Project Pri Due Active Age Description
153     asdf                   17s task1
154 153 asdf                    8s task2      

2 tasks
task> next pro:asdf

Pri Due Active Project ID  Description
               asdf    153 task1
               asdf    154 task2      

2 tasks
</pre>

To take blocked tasks off my active lists, I either filter them using "depend:" or sort them by "depend+" to place them after unblocked tasks in the list:

<pre>
task> list pro:asdf dep:

ID  Dep Project Pri Due Active Age Description
153     asdf                    1m task1

1 task
</pre>

Task 2 is filtered, as it should.  Now, I mark task1 as completed.

<pre>
task> 153 done
Completed task 153 'task1'.
Unblocked 154 'task2'.
Completed 1 task.
Project 'asdf' is 66% complete (1 of 3 tasks remaining).
</pre>

Why "1 of 3 tasks"?

Now, "list" command shows task 153 correctly:

<pre>
task> list pro:asdf    

ID  Dep Project Pri Due Active Age Description
153     asdf                    1m task2

1 task
</pre>

But nothing is shown when "dep:" filter is used.  This is a problem.

<pre>
task> list pro:asdf dep:
No matches.
task> 

</pre>

And edit command reveals that task 153 still has the dependency:

<pre>
1.  ID:                153
1.  UUID:              5bf1ed0d-3134-46e9-8aae-97fead1601ad
1.  Status:            Pending
1.  Mask:              
1.  iMask:             
  Project:           asdf
  Priority:          
1.  Separate the tags with spaces, like this: tag1 tag2
  Tags:              
  Description:       task2
  Created:           6/19/2012
  Started:           
  Ended:             
  Due:               
  Until:             
  Recur:             
  Wait until:        
  Parent:            
  Foreground color:  
  Background color:  
1.  Annotations look like this: <date> -- <text> and there can be any number of them.
1.  The ' -- ' separator between the date and text field should not be removed.
1.  A "blank slot" for adding an annotation follows for your convenience.
  Annotation:        6/19/2012 -- 
1.  Dependencies should be a comma-separated list of task IDs/UUIDs, with no spaces.
  Dependencies:      51bcf019-dc03-4ef1-ab8d-e66f320a543d
1.  End
</pre>

When "Dependencies:" line is removed, the task shows up in the "next" list as it should.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2012-06-19T23:30:42Z
Modified: 2014-02-09T02:04:40Z
```

**Comment by taskwarrior** [comment]
_Arkady Grudzinsky on 2012-06-19T23:31:28Z says:_ 

 I'm using task 2.0.0
<pre>
task> version

task 2.0.0 built for linux-lua
</pre>

**Comment by taskwarrior** [comment]
_Arkady Grudzinsky on 2012-06-19T23:35:55Z says:_ 

 When I print a list of blocked tasks, the list contains a handful of tasks which are not greyed out and there is nothing in "depend" column:

<pre>
task> blocked pro:asdf

ID  Deps Project Pri Due Active Age Description
153      asdf                   29m task2

1 task
</pre>

As you see, "Deps" column is empty, but the task is still considered "blocked"

**Comment by taskwarrior** [comment]
_Arkady Grudzinsky on 2012-06-19T23:45:01Z says:_ 

 One other observation is that while task1 is "alive", "Dependencies:" of task2 shows the task ID.  After task1 is marked "done", "Dependencies:" line shows UUID - the long string.

**Comment by taskwarrior** [comment]
_Arkady Grudzinsky on 2012-06-20T00:03:01Z says:_ 

 I think, when task is unblocked, the "Dependencies:" string should be left empty, and not replaced with UUID of completed task.  As-is, I do not see a way to list the tasks affected by this bug.  I must do "task blocked" and then remove "Dependencies:" line using "task edit" for each task I find in the list.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2012-07-07T20:49:36Z says:_ 

 There are a couple of problems here.  First note that we do not want to remove a dependency, just because a task was completed.  We want the data relationship to remain intact.  The problem you are encountering most is that "blocked" means "has a dependency", not "has an dependency that is not completed".  The latter is an expensive calculation which is just being ignored right now.

We need to think hard about this one, and find a way to make those calculations cheaper, without the need to parse all current and old tasks in order to do that.

Thanks for the report.  No solution yet.

**Comment by taskwarrior** [comment]
_Arkady Grudzinsky on 2012-07-08T04:19:22Z says:_ 

 This is a major issue for me.  To the point that I switched to GTG for the time being.  When a task is blocked by another task, I mark the dependency to get it out of the way with a full expectation that it will pop up at the top of the list when dependency is resolved.  This behavior is far more important for me than keeping record of the dependencies.  I rarely dig into completed tasks. I started losing track of my projects because of this, and had to search for these hidden tasks manually to delete dependencies.  I don't know of a good way to find these tasks automatically.

**Comment by taskwarrior** [comment]
_Johannes Schlatow on 2012-07-08T08:34:11Z says:_ 

 Paul Beckingham wrote:

> We need to think hard about this one, and find a way to make those calculations cheaper, without the need to parse all current and old tasks in order to do that.

Isn't this bounded by the number of pending tasks? I mean in order to calculate whether the dependency is completed or not, we "only" have to look up the task in pending.data.

As a compromise we could perhabps introduce a dedicated filter for "(in)active dependencies" in addition to (at)depends.none@ and (at)depends.any@ so that the user can decide if she can afford those expensive calculations.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2012-07-08T13:23:44Z says:_ 

 > Isn't this bounded by the number of pending tasks? I mean in order to calculate whether the dependency is completed or not, we "only" have to look up the task in pending.data.

That's correct.  I misspoke earlier wen I said it needed to scan old + new.  Just pending.data is enough to satisfy the query.  Urgency is another expensive calculation, and is one that is not performed unless needed, and when it is done, the result is cached.  I think we could handle this the same say.  So after sleeping on this, I'm more confident that we can determine "blocked" accurately:  it just requires that we follow all dependency chains until the end, where the end is a pending task or one that is not in pending.

> As a compromise we could perhabps introduce a dedicated filter for "(in)active dependencies" in addition to (at)depends.none@ and (at)depends.any@ so that the user can decide if she can afford those expensive calculations.

That's a possibility, but I'm not sure we need to go that far.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2012-07-08T21:51:42Z says:_ 

 Good news: this is implemented.  It's better than the old semi-approximate method, and faster too.  Thanks for the idea Johannes, and Arkady for communicating the severity.  The bad news is that it breaks a lot of tests, but those are tests that tolerated approximations of blocked/blocking.  This will make it into 2.1.0.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2012-07-09T05:31:30Z says:_ 

 Fixing this improves the urgency calculation, the next report, and the color rules.  Thanks Arkady.

**Comment by taskwarrior** [comment]
_Louis-Claude Canon on 2012-07-09T08:37:20Z says:_ 

 Although there may be performance improvements (and with cleaner code), I wonder which bug (if any) was solved. I got some issues with the colours and task initialisation that make some tests failing, but the provided patch seems to solve everything.

I think that the initial issue remains: to print only blocked tasks when running the "task blocked" report (and vice versa for unblocked). I think it would be nice if derived attributes like "is_blocked" and "is_blocking" could be used to filter tasks.

**Comment by taskwarrior** [comment]
_Kathryn Andersen on 2012-08-07T03:11:07Z says:_ 

 Did this fix make it into 2.10 or not? I'm puzzled, because I'm using taskwarrior 2.11, and this issue is does not appear to be fixed.

Action:
<pre>
task add pri:H foo
task add pri:H bar
(say these are tasks 249 and 250)
task 250 modify depends:249
</pre>

Doing a "task next" shows task 249 and not task 250 - this is correct.

<pre>
task 249 done
</pre>

This correctly declares that task 250 ("bar") is unblocked.

Doing a "task next" no longer shows "foo" - this is correct.

However, doing a "task next" does not show "bar" - this is NOT correct.
Because "bar" is no longer blocked by "foo", it SHOULD appear in the "next" report. And yet it does not appear.

It does not appear because of a two things: (a) when "foo" is completed, "bar" is altered to depend on the UUID of "foo" rather than the task-id of "foo" as it was before; (b) the "next" report checks for no-dependencies rather than somehow checking for not-blocked.

One or both of these things needs to be altered. Either add a specific test for "not-blocked" that the "next" report can use, or make it so that when a task is completed, all the tasks that used to depend on it no longer depend on it.
It really does not make any sense to depend on a task which is completed.

**Comment by taskwarrior** [comment]
_Arkady Grudzinsky on 2012-08-07T18:02:57Z says:_ 

 Something is still not right.  The task that was previously blocked still shows up on the "blocked" list after the blocker was marked as done.  I tested it in brand new set-up of task 2.1.1:
<pre>
task> add task1
Created task 1.
task> add task2
Created task 2.
task> next

ID Project Pri Due A Age Urgency Description
 1                   11s       0 task1
 2                    8s       0 task2      

2 tasks
task> 2 mod dep:1
Modifying task 2 'task2'.
Modified 1 task.
task> next

ID Project Pri Due A Age Urgency Description
 1                   35s       8 task1      
 2                   32s      -5 task2      

2 tasks
task> blocked

ID Deps Project Pri Due Active Age Description
 2 1                           44s task2      

1 task
task> unblocked

ID Deps Project Pri Due Active Age Description
 1                             50s task1      

1 task
task> 1 done
Completed task 1 'task1'.
Unblocked 2 'task2'.
Completed 1 task.
task> next

ID Project Pri Due A Age Urgency Description
 1                    1m       0 task2

1 task
</pre>
Now, this is the mystery:
<pre>
task> blocked

ID Deps Project Pri Due Active Age Description
 1                              1m task2

1 task
task> unblocked
No matches.
task> 

</pre>

**Comment by taskwarrior** [comment]
_Louis-Claude Canon on 2012-08-08T10:56:39Z says:_ 

 The behaviour you describe is related to issue #838, which is still not fixed.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2012-08-08T11:31:56Z says:_ 

 I have a fix for #838 coming shortly.

**Comment by pbeckingham** [comment]
[TW-527_0001-Bug-1022.patch.txt](https://github.com/GothenburgBitFactory/taskwarrior/files/1756155/TW-527_0001-Bug-1022.patch.txt)
